File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        n = len(nums)
        nums.sort()

        def countpair(target):
            count = 0
            first = 0
            last = n - 1

            while first < last:
                if nums[first] + nums[last] <= target:
                    count += (last - first)
                    first += 1
                else:
                    last -= 1
            return count

        totalupper = countpair(upper)
        totallower = countpair(lower - 1)
        ans = totalupper - totallower

        return ans

File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.js
/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {number}
 */
var countFairPairs = function (nums, lower, upper) {
    nums.sort((a, b) => a - b);
    return countPairs(nums, upper) - countPairs(nums, lower - 1);
};

function countPairs(nums, target) {
    let count = 0;
    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        if (nums[left] + nums[right] > target) {
            right--;
        } else {
            count += right - left;
            left++;
        }
    }

    return count;
}